File: functional-programming-in-python/exercise02.py
employees = [
    {"identity": "1", "fullname": "jack bauer", "salary": "10000", "iban": "tr1", "department": "IT",
     "birthYear": 1964},
    {"identity": "2", "fullname": "kate austen", "salary": "20000", "iban": "tr2", "department": "SALES",
     "birthYear": 1984},
    {"identity": "3", "fullname": "james sawyer", "salary": "30000", "iban": "tr3", "department": "HR",
     "birthYear": 1982},
    {"identity": "4", "fullname": "jin kwon", "salary": "40000", "iban": "tr4", "department": "IT", "birthYear": 1988},
    {"identity": "5", "fullname": "sun kwon", "salary": "50000", "iban": "tr5", "department": "FINANCE",
     "birthYear": 1987}
]


def order_by_salary(emp):
    return emp["salary"]


def order_by_age(emp):
    return 2022 - emp["birthYear"]


def order_by_department(emp):
    return emp["department"]


def order_by(field): # higher-order function
    return lambda e: e[field]


for employee in employees:
    print(employee)
employees.sort(key=order_by("birthYear"), reverse=False)  # higher-order function
for employee in employees:
    print(employee)

File: module03/exercise02.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

#include "employee.h"

template<int Age>
bool older_than(const employee &emp) {
    return (2023 - emp.getBirthYear()) > Age;
}

struct OlderThan {

    auto
    operator()(const employee &emp) -> bool {
        return (2023 - emp.getBirthYear()) > limit;
    }

    explicit OlderThan(int limit=42) : limit(limit) {}

private:
        int limit;
};

template <int Limit>
struct OlderThan2 {

    auto
    operator()(const employee &emp) -> bool {
        return (2023 - emp.getBirthYear()) > Limit;
    }

};
int main() {
    vector<employee> employees{
            {"kate",   "austen", employee::department_t::it,      employee::gender_t::female, 200'000, "tr1", 1986},
            {"james",  "sawyer", employee::department_t::hr,      employee::gender_t::male,   100'000, "tr2", 1976},
            {"juliet", "burke",  employee::department_t::sales,   employee::gender_t::female, 100'000, "tr3", 1996},
            {"john",   "locke",  employee::department_t::finance, employee::gender_t::male,   200'000, "tr4", 1956},
            {"ben",    "linus",  employee::department_t::it,      employee::gender_t::male,   100'000, "tr5", 1966}
    };
    auto counter = count_if(employees.begin(), employees.end(), older_than < 50 > ); // HoF
    cout << counter << endl;
    counter = count_if(employees.begin(), employees.end(), older_than < 40 > ); // HoF
    cout << counter << endl;
    counter = count_if(employees.begin(), employees.end(), OlderThan(30) ); // HoF
    cout << counter << endl;
    counter = count_if(employees.begin(), employees.end(), OlderThan2<30>() ); // HoF
    cout << counter << endl;
    auto limit = 30;
    counter = count_if(employees.begin(), employees.end(), [&limit](const employee &emp) -> bool {
        return (2023 - emp.getBirthYear()) > limit;
    } ); // HoF
    cout << counter << endl;

    return 0;
}
